### PR TITLE
[FLINK-14816] Add thread dump feature for taskmanager

### PR DESCRIPTION
**Pull request by tillrohrmann**
Extension of #10228.

### CODE REVIEW

**Comment by lamber-ken** [comment]
hi @tillrohrmann, thanks for doing this, 💯

**Comment by tillrohrmann** [comment]
Thanks for the pointer @lamber-ken. I will update the PR to fix the `RestAPIStabilityTest`.

**Comment by tillrohrmann** [comment]
cc @zentol: The `TaskManagerThreadDumpHandler` now returns a `ThreadDumpInfo` which contains an array of `ThreadDumpInfo.ThreadInfo`. `ThreadDumpInfo.ThreadInfo` contains the name of the thread and the stringified thread info. This is kind of the compromise I found wrt our offline discussion.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 1]
```diff
@@ -1675,15 +1703,12 @@ private boolean isJobManagerConnectionValid(JobID jobId, JobMasterId jobMasterId
...
 				final File file = new File(filePath);
 				if (file.exists()) {
-					final TransientBlobCache transientBlobService = blobCacheService.getTransientBlobService();
-					final TransientBlobKey transientBlobKey;
-					try (FileInputStream fileInputStream = new FileInputStream(file)) {
-						transientBlobKey = transientBlobService.putTransient(fileInputStream);
-					} catch (IOException e) {
+					try {
+						return putTransientBlobStream(new FileInputStream(file), fileTag).get();
+					} catch (Exception e) {
```
Comment (lamber-ken) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1708]: hi @tillrohrmann, since doesn't make thread dump as stream in your design right now, we can revert it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 2]
```diff
@@ -1012,10 +1016,34 @@ public void disconnectResourceManager(Exception cause) {
...
+
+		try {
+			transientBlobKey = transientBlobService.putTransient(inputStream);
+		} catch (IOException e) {
+			log.debug("Could not upload file {}.", fileTag, e);
+			return FutureUtils.completedExceptionally(new FlinkException("Could not upload file " + fileTag + '.', e));
+		}
+		return CompletableFuture.completedFuture(transientBlobKey);
+	}
+
```
Comment (lamber-ken) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1046]: hi @tillrohrmann, since doesn't make thread dump as stream in your design right now, we can revert it.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1046]: I think this change is still good since it breaks up a longer method into smaller units.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java [hunk 3]
```diff
@@ -22,6 +22,13 @@
  * Different file types to request from the {@link TaskExecutor}.
  */
 public enum FileType {
+	/**
+	 * the log file type for taskmanager
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java:26]: this seems unrelated?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 4]
```diff
@@ -1012,10 +1016,34 @@ public void disconnectResourceManager(Exception cause) {
 		}
 	}
 
+	@Override
+	public CompletableFuture<ThreadDumpInfo> requestThreadDump(Time timeout) {
+		final Collection<ThreadInfo> threadDump = JvmUtils.createThreadDump();
+
+		final Collection<ThreadDumpInfo.ThreadInfo> threadInfos = threadDump.stream()
+			.map(threadInfo -> ThreadDumpInfo.ThreadInfo.create(threadInfo.getThreadName(), threadInfo.toString()))
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1024]: I'm wondering whether this conversion should really be done on the TaskExecutor.
I'd prefer if conversions to REST responses are done by the handler just to separate things a bit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * Class containing thread dump information.
+ */
+public class ThreadDumpInfo implements ResponseBody, Serializable {
+	private static final long serialVersionUID = 1L;
+
+	public static final String FIELD_NAME_THREAD_INFOS = "threadInfos";
+
+	@JsonProperty(FIELD_NAME_THREAD_INFOS)
+	private final Collection<ThreadInfo> threadInfos;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java:39]: Why not a map thread_name -> thread_info? I think this would be more generally usable.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1046]: They still belong into a separate commit though.

Comment (lamber-ken) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1046]: > I think this change is still good since it breaks up a longer method into smaller units.

ok

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java:26]: Somewhat. But I believe that the extra effort splitting this change into a separate commit won't pay off. Hence, I would like to keep it as is.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1024]: As soon as the RestResponse and the internal cluster message differ I will introduce a separate the classes. As long as they are identical I think it is fine to use the same class here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java:39]: The main reason was the duplicated thread name information as a key and as part of the `ThreadInfo` class. But I guess this is premature optimization. I will change it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1046]: Technically yes but I think it is ok to not twiddle with this commit because the benefit would be marginal.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java:39]: Actually, I think whether it is stored in a map or not is an implementation detail. Moreover, we don't offer a method `getThreadInfo(String threadName)` yet. Hence, there is actually no reason to store them internally as a map. Once the requirement changes, I'm happy to adapt the class.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. I answered all your comments. Please take another look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java (hunk 5 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java:39]: I wasn't talking about changing the internal storage, but to expose it as a map instead (i.e., changing the JOSN representation)

Assuming a user is interested in a particular thread, then with a map one can do `threadDump['myThreadName']`, instead of having to iterate over the array.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java:26]: Per the [code-style guide](https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#separate-refactoring-cleanup-and-independent-changes):
> Pull Requests must put cleanup, refactoring, and core changes into separate commits.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/FileType.java:26]: You are right Chesnay. I will refactor the changes into separate commits in order to follow the code style guide.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java:39]: Good point. I will take another look.

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java:39]: After having looked at it again, I still believe that the response should be a collection of `ThreadInfos`. If the user needs to do a named lookup, then she can still create a map from the collection. It should not be the responsibility response class to present the elements in a certain order. Moreover, we don't have this requirement at the moment. Hence, I won't do the proposed change.

**Comment by tillrohrmann** [comment]
Thanks for the review @lamber-ken.

@zentol I've split the unrelated changes into separate commits. I won't change the JSON format though. I will merge this PR once CI gives green light.
